doc/api/packages.md

```diff
@@ -274,40 +275,31 @@ a project that previous exported `main`, `lib`,
 }
 ```
 
-Alternatively a project could choose to export entire folders:
+Alternatively a project could choose to export entire folders using export
+patterns:
 
 ```json
 {
   "name": "my-mod",
   "exports": {
     ".": "./lib/index.js",
     "./lib": "./lib/index.js",
-    "./lib/*": "./lib/*.js",
+    "./lib/*": "./lib/*",
```

Comment:
Entry points shouldn’t contain extensions; won’t this change mean consumers have to provide the .js?

Reply:
See bullet point two - one of the changes in this PR is to encourage explicit entry extensions, since it enables compressible import maps via `./` mappings in the map, which would otherwise require explicit mappings.

It's certainly something that may have arguments either way though, and if there is resistance to the change I could remove this aspect of the PR, although it is one of the motivating reasons in trying to guide the ecosystem towards a place where import maps don't end up bloated.

Reply:
One of the things that came up in the collaborator summit is that Node needs first-class support for import maps, so +1 on starting to move us in that direction.

Reply:
That contradicts the longstanding ecosystem preference, and i don’t think it belongs in node’s docs.

Import maps from node_modules are useless unless you a) serve node_modules, a horrifically bad practice, and b) have a 100% ESM tree, which virtually nobody does, so i don’t think it’s valuable to artificially push the ecosystem in that direction.

Reply:
Having `"exports"` compatible with import maps lets users generate an import maps that use e.g. `unpkg.com` so you could start using your JS app (with Deno, a web browser, Node.js once it has import maps support, etc.) without needing a node_modules folder. My interpretation of things is not that Node.js is artificially pushing the ecosystem to ESM, but rather the ecosystem is organically pushing Node.js to ESM and interoperability with other tools of the ecosystem, and import maps are part of this.

Alternatively, we could give an example of an import maps, and illustrate how export pattern, while supported by Node.js, may not be desirable because it makes the process of generating an import maps for your module more complicated and error-prone. That would let users choose one or the other depending on wether import maps support matters to their use-case.

So given this `package.json`, we can use the following import maps:

```json
{
  "imports": {
    "my-mod": "https://unpkg.com/my-mod/lib/index.js",
    "my-mod/lib": "https://unpkg.com/my-mod/lib/index.js",
    "my-mod/lib/": "https://unpkg.com/my-mod/lib/"
  }
}
```

If the `package.json` uses export pattern, it would look something like:

```json
{
  "imports": {
    "my-mod": "https://unpkg.com/my-mod/lib/index.js",
    "my-mod/lib": "https://unpkg.com/my-mod/lib/index.js",
    "my-mod/lib/index": "https://unpkg.com/my-mod/lib/index.js",
    "my-mod/lib/my-mod": "https://unpkg.com/my-mod/lib/my-mod.js",
     …
    "my-mod/lib/yetAnotherFile": "https://unpkg.com/my-mod/lib/yetAnotherFile.js",
  }
}
```

BTW, `my-mod` is not a very good package name, it blurs the line between a package and a module which is very confusing.

Reply:
Any such import map generating tool can do so whether extensions or used or not. It is a fallacy that this has any meaningful impact on import map generation imo.

Reply:
The second example @aduh95 provides above clearly has more entries than the first. If we had `*` patterns in import maps, sure it wouldn't be an issue, but import maps have web compat at this point such that their matching syntax is locked and unlikely to change.

Reply:
Are they shipped in a non chrome browser?

Reply:
Yes, under a flag in Firefox.

Reply:
+1 on renaming `my-mod`; maybe `package`? Also the use of `main.js` might be confusing since you also discuss `"main"`; maybe rename that to `entry.js`?

Reply:
I've renamed `my-mod` to `my-package` and replaced usage of `main.m?js` with `index.m?js`, let me know how that looks.

Reply:
"under a flag" isn't part of the web, so that means they're just shipping in chrome.

Reply:
Besides the web, import maps are a core feature of Deno: https://deno.land/manual/linking_to_external_code/import_maps. We’re also trying to improve interoperability between server-side JavaScript runtimes.

Reply:
Sure, that’s all fine, extensionless entry points however remain dominant, idiomatic, and supported, so I’m not sure why “a feature node doesn’t support” would have more weight than that?

Reply:
Because the example here is explicitly about supporting backwards-compatibility, I've updated the example to use the pattern of exporting support for exports both with and without extensions. This should resolve this concern I believe.